scripts/Diffraction/single_crystal/base_sx.py

```diff
@@ -614,6 +618,155 @@ def remove_non_integrated_peaks(peaks):
+                        wsMD, pk, peak_shape, nbins_max, extent, frame_to_peak_centre_attr
+                    )
+
+                    fig, axes = subplots(1, 3, figsize=(12, 4), subplot_kw={"projection": "mantid"})
+                    for iax, ax in enumerate(axes):
+                        dims = list(range(3))
+                        dims.pop(iax)
+                        # plot slice
+                        im = ax.imshow(ws_cut.getSignalArray().sum(axis=iax)[:, ::-1].T)  # reshape to match sliceviewer
+                        im.set_extent([-1, 1, -1, 1])  # so that ellipsoid is a circle
+                        if log_norm:
+                            im.set_norm(LogNorm())
+                        # plot peak position
+                        ax.plot(0, 0, "xr")
+                        # plot peak representation (circle given extents)
+                        patch = Circle((0, 0), radii[imax] / box_lengths[imax], facecolor="none", edgecolor="r", ls="--")
+                        ax.add_patch(patch)
+                        # add background shell
+                        if np.all(bg_outer_radii > 1e-10):
+                            patch = Circle(
+                                (0, 0), bg_outer_radii[imax] / box_lengths[imax], facecolor="none", edgecolor=3 * [0.7]
+                            )  # outer radius
+                            ax.add_patch(patch)
+                            patch = Circle(
+                                (0, 0), bg_inner_radii[imax] / box_lengths[imax], facecolor="none", edgecolor=3 * [0.7], ls="--"
+                            )  # inner radius
+                            ax.add_patch(patch)
+                        # format axes
+                        ax.set_xlim(-1, 1)
+                        ax.set_ylim(-1, 1)
+                        ax.set_aspect("equal")
+                        ax.set_xlabel(ws_cut.getDimension(dims[0]).name)
+                        ax.set_ylabel(ws_cut.getDimension(dims[1]).name)
+                    fig.suptitle(
+                        f"{ipk} ({','.join(str(np.round(pk.getHKL(), 2))[1:-1].split())})"
+                        f"  $I/\\sigma$={np.round(pk.getIntensityOverSigma(), 2)}\n"
+                        rf"$\lambda$={np.round(pk.getWavelength(), 2)} $\AA$; "
+                        rf"$2\theta={np.round(np.degrees(pk.getScattering()), 1)}^\circ$; "
+                        rf"d={np.round(pk.getDSpacing(), 2)} $\AA$"
+                    )
+                    fig.tight_layout()
+                    pdf.savefig(fig)
+                    close(fig)
+
+                if ws_cut is not None:
+                    mantid.DeleteWorkspace(ws_cut)
+        except OSError:
+            raise RuntimeError(
+                f"OutputFile ({filename}) could not be opened - please check it is not open by "
+                f"another programme and that the user has permission to write to that directory."
+            )
```

Comment:
This might be tricky - I noticed a similar try-except statement is used in other places such as `IntegratePeaksSkew`and `IntegratePeaksShoeboxTOF`. I was wondering if it would be worth the effort moving the function to a common location so it can be reused and save you time in the future?

Reply:
Good spot, there's actually a lot of code in `IntegratePeaksSkew` that is used in the other algorithms I wrote later (`IntegratePeaksShoeboxTOF` and `FindSXPeaksConvolve`) - in particular the `PeakData` class. I think all of the common code used in these algorithms belongs in a separate helper algorithm (I think you have something like this in quasibayes?).  Would you be OK for me to make a maintenance issue for this?

Reply:
Yes no problem!

Reply:
Have made issue here #36989